### PR TITLE
Fixed a long waiting issue about file autto change detection

### PR DESCRIPTION
**Pull request by SinghRajenM**
@donho This mega PR fixes below defects:
#1087
#3349
#3333
#3528
#3138
#3934
#3160
#1822
#745
#3599
#4329
#3971
#4592

Following defects are no longer valid, as these cases will not occur
#5386
#4189
#1995
#2867
#3306
#2776
#2724
#2405
#4072
#4747

**_The list is huge and there are many more similar defects. I will keep on adding them here._**

**Edit:** _Below screenshot is no longer valid as changes were made to accommodate review comments._
image

### CODE REVIEW

**Comment by SinghRajenM** [comment]
Fixes:
#1116
#2082

No longer valid:
#4180

**Comment by sasumner** [comment]
Big big thumbs up...if if gets accepted/integrated.  Otherwise, meh. :)

**Comment by LordNyriox** [comment]
@SinghRajenM:  **Grammar Warning:** 

Though not necessarily incorrect in English, the article in "all the tabs" suggests that there is a *specific set* of tabs being referenced—rather than a generic global definition. 

Perhaps you can change the option text to read "for all tabs" instead?

**Comment by LordNyriox** [comment]
@SinghRajenM:  **Grammar Warning:** 

Though not necessarily incorrect in English, the article in "`all the tabs`" suggests that there is a *specific set* of tabs being referenced—rather than a generic global definition. 

Perhaps you can change the option text to read "`for all tabs`" instead? 

**P.S.**  Also note that there is no article in the label for the newly added option "for current tab".  Please either include "the" in both labels, or neither of them.

**Comment by LordNyriox** [comment]
@SinghRajenM:  **Grammar Warning:** 

Though not necessarily incorrect in English, the article in "`all the tabs`" suggests that there is a *specific set* of tabs being referenced—rather than a generic global definition. 

Perhaps you can change the option text to read "`for all tabs`" instead? 

**P.S.**  Also note that there is no article in the label for the newly added option "`for current tab`".  Please either include "`the`" in both labels, or neither of them.

**Comment by sasumner** [comment]
@LordNyriox I see "for all the tabs" in the image above; maybe the image was changed AFTER your comment?

**Comment by SinghRajenM** [comment]
@sasumner No. No change till now.

As per @LordNyriox 
either remove "the" from "Enable for all ~~the~~ tabs" 
or add "the" in "Enable only for **the** current tab".

**Comment by sasumner** [comment]
IMO the best thing would be to drop "the" in both.

**Comment by wqweto** [comment]
Can you please merge #5273 in this PR (a simple 5 lines change) as it's totally ignored so far by the maintainer.

The proposed fix allows to **undo** the automatic/manual file reloading. We are humans making mistakes and can confirm file reload even though we don't want to lose our unsaved modifications. Let's give a chance to undo such mistake!

**Comment by SinghRajenM** [comment]
> IMO the best thing would be to drop "the" in both.

Done now.

>Can you please merge #5273 in this PR (a simple 5 lines change) as it's totally ignored so far by the maintainer.

@wqweto It is better to keep #5273 separate to adhere [Contributing Policy](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#guidelines-for-pull-requests). Also as I can @donho  has assigned #5273 to himself. Once he will get time, he will review and take action. I would say, don't worry, it is not ignored, just taking time.

**Comment by Tatsuran** [comment]
@SinghRajenM thank you so much. Waiting for update with this fix a lot, switching tabs issue driving me nuts

**Comment by donho** [comment]
@SinghRajenM  What is the new behaviour? A detail description will help to integrate your PR.

**Comment by SinghRajenM** [comment]
@donho 
Please find the description as below - 
1. **If new checkbox is not checked, then old behavior is retained. No change at all.** It will help those users who want to detect changes for all the opened files. So it is retained as it is.
2. If new checkbox is checked, then
     a. If file change will be detected only for active tab. So user will see only one messagebox for reload. If file other than active tab is modified, then user will be see reload message only when s/he activates that particular tab.
    b. Other two options are same as they were working before for all the tab (but now will work for current tab too.). Like if user check "update silently", then the active will be updated silently without showing messagebox. Same for "Go to end" and combination of both.

From UI perspective:
1. "Update silently" and "Scroll to the end after update" will be enabled only if either one of options (Enable for all tabs or Enable for current tab) is selected.
2. At a given time, user can choose "Enable for all tabs" or "Enable for current tab" or None of them.

File: PowerEditor/src/NppBigSwitch.cpp [hunk 1]
```diff
@@ -1621,15 +1617,15 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
 
 		case NPPM_INTERNAL_ENABLECHECKDOCOPT:
 		{
-			NppGUI & nppgui = const_cast<NppGUI &>((pNppParam->getNppGUI()));
+			NppGUI& nppgui = const_cast<NppGUI&>((pNppParam->getNppGUI()));
```
Comment (SinghRajenM) [inline PowerEditor/src/NppBigSwitch.cpp]: @donho 
**Additional note:**
I feel "Doc Monitor" plugin developer needs to intimated. With this current implementation, new functionalities can not be enabled. So this plugin developer should be informed and we may need to provide additional support to him/her if s/he is actively updating plugin.

**Comment by donho** [comment]
Good job.
It seems to me the new behaviour is better than the old one.
Let's remove the old behaviour and make new one as standard behaviour.

**Comment by SinghRajenM** [comment]
@donho Updated the PR as per review comments.

**Comment by bahusoid** [comment]
Please confirm that Save All operation does nothing for unchanged tabs, so it's safe to click it even if some tabs were changed externally.

And what about "Replace All in All Opened Documents" operation? Does it mean that with this new behavior replace will be silently applied to old document and external changes are lost on Save All?

**Comment by SinghRajenM** [comment]
@bahusoid 
> Please confirm that Save All operation does nothing for unchanged tabs, so it's safe to click it even if some tabs were changed externally.

Yes, save all will not impact unchanged tab as well as externally modified file.


>And what about "Replace All in All Opened Documents" operation?

This is behavioral decision here. 
**If only file changes detection is enabled**, "Replace All in All Opened Documents" or "File all in All opened Doc" will be performed on current state of all the tab. That means if something is externally modified will not be considered. **But, if "Update silently"** is also enabled then externally modified changes will be considered during "Replace All in All Opened Documents" and "File all in All opened Doc" operations. So it is all user's choice.

**Comment by bahusoid** [comment]
> This is behavioral decision here.
>If only file changes detection is enabled, "Replace All in All Opened Documents" or "File all in All opened Doc" will be performed on current state of all the tab. 

But it's not desired behavior. I want to be informed on changes. And when I do "Replace All" I want to be sure that this operation is done on up to date files.
With this pr it's either silent update or  silent replace on outdated file  - not good.

**Comment by donho** [comment]
@SinghRajenM What @bahusoid pointed out is an expected behaviour but hard to implement.
To prevent from users' complain about unexpected side effect of the new behaviour, I will keep both old and new behaviours. So users can switch back the old one if they want. You don't need to do anything on this PR, I kept your previous one on my local branch.

Thank you for this helpful improvement.

**Comment by SinghRajenM** [comment]
@donho 

> @SinghRajenM What @bahusoid pointed out is an expected behaviour but hard to implement.

I think I found the solution for this issue as well. Likely one line solution 😜 

``` 
bool Notepad_plus::replaceInOpenedFiles() {
......
.....
if (pBuf->isReadOnly())
				continue;

//These two line are newly added
			if(!pBuf->isDirty())
				pBuf->checkFileState();
```

Say there are 5 tabs opened.
Tab 1: Saved
Tab 2: Externally modified
Tab 3: Externally modified
Tab 4: A dirty tab (may be saved, but dirty)
Tab 5: A saved dirty tab, but it is modified outside as well.

Currently, I'm on Tab 1. I do replace all in opened tab.

Result:
**Tab 1**: replacement will happen without any issue.
**Tab 2**: It will ask for confirmation to reload
     a) if user says yes to reload, then file will be reloaded and then replacement will happen.
     b) if user says no to reload, then replacement will happen for old data
**Tab 3**: Unless user come out of tab 2, it will be waiting for its turn. Once tab 2 operation is complete, replacement will happen for this tab as per 2.a and 2.b
**Tab 4**: After tab 3, replacement will happen on this tab directly.
**Tab 5**: After tab 4, replacement will happen for this tab as well _but with old data_.

Please note: User can do <kbd>Ctrl</kbd> + <kbd>z</kbd> for all the 5 tabs.

I feel this is expected. I feel this can be applied to find operation as well. I will check in sometime today.

**Comment by bahusoid** [comment]
@SinghRajenM Sounds good! But that makes  "Do for all" request relevant again :)

**Pull request closed without merging**